toolchain/lex/tokenized_buffer.cpp

```diff
@@ -556,14 +557,14 @@ class [[clang::internal_linkage]] TokenizedBuffer::Lexer {
     }
 
     if (literal->is_terminated()) {
-      auto token =
-          buffer_.AddToken({.kind = TokenKind::StringLiteral,
-                            .token_line = string_line,
-                            .column = string_column,
-                            .literal_index = static_cast<int32_t>(
-                                buffer_.literal_string_storage_.size())});
-      buffer_.literal_string_storage_.push_back(
-          literal->ComputeValue(emitter_));
+      buffer_.computed_strings_.push_back(
+          std::make_unique<std::string>(literal->ComputeValue(emitter_)));
```

Comment:
This allocation seems to be redundant.

Reply:
Is this about the StringMap? Please see my comments there, I think this isn't redundant if we stop using StringMap.

Reply:
This was primarily about the `StringMap`, but even with the `DenseMap` this is performing an extra allocation for longer strings, and will overallocate for shorter strings that fit easily into the small-string buffer (which I expect to be a common case.) Maybe consider something like:
```suggestion
      std::string value = literal->ComputeValue(emitter_);
      auto storage = std::make_unique<char[]>(value.size());
      std::copy(value.begin(), value.end(), storage.get());
      buffer_.computed_strings_.push_back(std::move(storage));
```
... or just add a TODO so we don't forget about this in the longer term.

(Having said all of that, I think the most common case will be where we can reuse the source buffer to provide the contents of the string -- if it contains no escapes and (if it's multiline) has no indentation -- and we're not optimizing for that case yet, so perhaps that's the more important TODO here!)

Reply:
I'm going to make a TODO here, mainly to break this out. I thought though that this would be moved, without resulting in an extra allocation. My thought regardless is that we could have ComputeValue return the type we want though, rather than doing logic here (avoiding the `std::copy`), but I'd like to do that separately.